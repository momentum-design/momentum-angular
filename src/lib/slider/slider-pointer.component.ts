/** @component slider-pointer */

import {
  Component,
  ElementRef,
  EventEmitter,
  HostBinding,
  Inject,
  Input,
  Output,
} from '@angular/core';
import { DOCUMENT } from '@angular/common';

export interface SliderPointerChange {
  from?: number;
  to?: number;
  direction: number;
  isKeyBoard: boolean;
}

@Component({
  selector: 'cui-slider-pointer',
  template: ``,
  host: {
    class: 'cui-slider__pointer',
    role: 'button',
    tabindex: '0',
    '(keydown)': 'onKeydown($event)',
    '(mousedown)': 'onMousedown($event)',
    '(touchstart)': 'onTouchstart($event)',
    '(touchmove)': 'onTouchmove($event)',
  },
})
export class SliderPointerComponent {
  private _previousPosition: number;
  private _document: Document;

  /** @option Set Slider Pointer's position | 0 */
  @Input() position: number = 0;

  /** Event emitted when user moves the Slider Pointer */
  @Output() readonly move: EventEmitter<SliderPointerChange> = new EventEmitter<
    SliderPointerChange
  >();

  @HostBinding('style.left.%') get left(): number {
    return this.position;
  }

  constructor(
    private elementRef: ElementRef,
    @Inject(DOCUMENT) _document: any
  ) {
    this._document = _document;
  }

  getDirections(currentPos: number) {
    if (currentPos > this._previousPosition) {
      return 1;
    }
    return -1;
  }

  onKeydown(event: KeyboardEvent) {
    switch (event.code) {
      case 'ArrowUp':
      case 'ArrowRight':
        this.move.emit({
          direction: 1,
          isKeyBoard: true,
        });
        event.preventDefault();
        break;

      case 'ArrowDown':
      case 'ArrowLeft':
        this.move.emit({
          direction: -1,
          isKeyBoard: true,
        });
        event.preventDefault();
        break;
    }
  }

  onMousedown(event: MouseEvent) {
    this._document.addEventListener('mousemove', this.onMousemove);
    this._document.addEventListener('mouseup', this.onMouseup);

    this._previousPosition = event.clientX;
  }

  onMousemove = (event: MouseEvent) => {
    const xPos = event.clientX;
    const direction = this.getDirections(xPos);

    this.move.emit({
      from: this.elementRef.nativeElement.getBoundingClientRect().left,
      to: xPos,
      direction: direction,
      isKeyBoard: false,
    });
  }

  onMouseup = (_: MouseEvent) => {
    this._document.removeEventListener('mousemove', this.onMousemove);
    this._document.removeEventListener('mouseup', this.onMouseup);
  }

  onTouchstart(event: TouchEvent) {
    this._previousPosition = event.touches[0].clientX;
  }

  onTouchmove(event: TouchEvent) {
    const xPos = event.touches[0].clientX;
    const direction = this.getDirections(xPos);

    this.move.emit({
      from: this.elementRef.nativeElement.getBoundingClientRect().left,
      to: xPos,
      direction: direction,
      isKeyBoard: false,
    });
  }
}
