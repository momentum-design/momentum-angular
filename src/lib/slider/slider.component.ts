import {
  AfterViewInit,
  ChangeDetectorRef,
  Component,
  ElementRef,
  EventEmitter,
  forwardRef,
  Input,
  OnInit,
  Output,
  Provider,
  ViewChild,
} from '@angular/core';
import { NG_VALUE_ACCESSOR, ControlValueAccessor } from '@angular/forms';
import { SliderPointerChange } from './slider-pointer.component';
import isObject from 'lodash-es/isObject';

// tslint:disable:no-use-before-declare
const CUI_SLIDER_VALUE_ACCESSOR: Provider = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => SliderComponent),
  multi: true,
};
// tslint:enable:no-use-before-declare

export interface SliderChange {
  high: number;
  low: number;
}

@Component({
  selector: 'cui-slider',
  template: `
    <span #sliderBar class="cui-slider__bar"></span>
    <span
      class="cui-slider__selection"
      [ngStyle]="{ 'left.%': selectionLeft, 'width.%': selectionWidth }"
    ></span>
    <cui-slider-pointer
      *ngIf="isObject(value)"
      [position]="sliderLowPosition"
      (move)="onSliderMove('sliderLow', $event)"
    ></cui-slider-pointer>
    <cui-slider-pointer
      [position]="sliderHighPosition"
      (move)="onSliderMove('sliderHigh', $event)"
    ></cui-slider-pointer>
    <div #ticksContainer>
      <span *ngFor="let tickValue of scale" class="cui-slider__hashlabel">{{
        (translateFn && translateFn(tickValue)) || tickValue
      }}</span>
    </div>
  `,
  host: {
    class: 'cui-slider',
    '[class.cui-slider--disabled]': 'disabled',
  },
  providers: [CUI_SLIDER_VALUE_ACCESSOR],
})
export class SliderComponent
  implements AfterViewInit, OnInit, ControlValueAccessor {
  /** @prop Determines if minimum pointer can cross over maximum pointer | false */
  @Input() canCross: boolean = false;
  /** @prop Set the attribute disabled to Slider | false */
  @Input() disabled: boolean = false;
  /** @prop Set the initial maximum value */
  @Input() max: number;
  /** @prop Set the initial minimum value | 0 */
  @Input() min: number = 0;
  /** @prop Set visual step measurement | 1 */
  @Input() step: number = 1;
  /** @prop Set increment of x-axis labels | 0 */
  @Input() tick: number = 0;
  /** @prop Function to compute layout of Slider | null */
  @Input() translateFn: Function | null;
  /** @prop Set either maximum pointer value or a combination of high and low pointer values | 0 */
  @Input()
  get value(): SliderChange | number {
    return this._value;
  }
  set value(value) {
    if (this._value !== value) {
      this._value = value;
      this.sliderLow =
        (isObject(this.value) && (this.value as SliderChange).low) || this.min;
      this.sliderHigh =
        (isObject(this.value) && (this.value as SliderChange).high) ||
        (this.value as number);
      this.returnCurrentValues();
      this.changeDetectorRef.markForCheck();
    }
  }
  /** Event emitted when change a pointer position */
  @Output() readonly change: EventEmitter<
    SliderChange | number
  > = new EventEmitter<SliderChange | number>();

  @ViewChild('sliderBar') private sliderBar: ElementRef;
  @ViewChild('ticksContainer') private ticksContainer: ElementRef;

  private _value: SliderChange | number = 0;
  private _sliderLow: number;
  private _sliderHigh: number;

  get sliderLow(): number {
    return this._sliderLow;
  }
  set sliderLow(value) {
    this._sliderLow = value;
  }
  get sliderHigh(): number {
    return this._sliderHigh;
  }
  set sliderHigh(value) {
    this._sliderHigh = value;
  }
  get sliderLowPosition() {
    return ((this.sliderLow - this.min) / (this.max - this.min)) * 100;
  }
  get sliderHighPosition() {
    return ((this.sliderHigh - this.min) / (this.max - this.min)) * 100;
  }

  scale: number[];
  selectionLeft: number;
  selectionWidth: number;
  isObject: Function;

  private _change(_: any) {}

  constructor(private changeDetectorRef: ChangeDetectorRef) {}

  ngOnInit() {
    if (!this.max) {
      throw new Error('Initial maximum(max) value is required');
    }
    this.isObject = isObject;
    this.scale = [this.min, this.max];

    this.getScale(this.min, this.max, this.tick);
    this.getSelectionWidth();
  }

  ngAfterViewInit() {
    this.setScalePos(this.scale, this.min, this.max);
  }

  writeValue(value: any): void {
    if (value !== null) {
      this.value = value;
    }
  }

  setDisabledState(disabled: boolean): void {
    this.disabled = disabled;
    this.changeDetectorRef.markForCheck();
  }

  registerOnChange(fn: any): void {
    this._change = fn;
  }

  registerOnTouched(_: any): void {}

  setScalePos(scale: number[], min: number, max: number) {
    this.ticksContainer.nativeElement
      .querySelectorAll('span')
      .forEach((child: HTMLElement, idx: number) => {
        const leftValue = `${((scale[idx] - min) / max) * 100}%`;
        child.style.left = `calc(${leftValue} - ${child.offsetWidth / 2}px)`;
      });
  }

  getScale(low: number = 0, high: number, tick: number) {
    if (this.tick > 0) {
      let value = high;
      const ticksArray: number[] = [high];
      while (value > 0 && value - tick >= low) {
        value -= tick;
        ticksArray.unshift(Math.abs(Math.round(value / tick) * tick));
      }
      this.scale = ticksArray;
    }
  }

  getSliderWidth() {
    return this.sliderBar.nativeElement.getBoundingClientRect().width;
  }

  getStepWidth() {
    const maxValue = this.max - this.min;
    const maxWidth = this.getSliderWidth();
    return (this.step / maxValue) * maxWidth;
  }

  getSteps(position: SliderPointerChange) {
    if (position.isKeyBoard) {
      return 1;
    }
    const diff =
      position.direction === 1
        ? position.to - position.from
        : position.from - position.to;
    if (diff < 0) {
      return 0;
    }
    const steps = diff / this.getStepWidth();
    return steps - Math.floor(steps) >= 0.5
      ? Math.ceil(steps)
      : Math.floor(steps);
  }

  getLimit(pointerKey: string, direction: number) {
    if (pointerKey === 'sliderLow') {
      return this.canCross
        ? direction === 1
          ? this.max
          : this.min
        : direction === 1
        ? this.sliderHigh
        : this.min;
    } else if (pointerKey === 'sliderHigh') {
      return this.canCross
        ? direction === 1
          ? this.max
          : this.min
        : direction === 1
        ? this.max
        : this.sliderLow;
    }
  }

  returnCurrentValues() {
    this.getSelectionWidth();
    const value = isObject(this.value)
      ? {
          low: Math.min(this.sliderHigh, this.sliderLow),
          high: Math.max(this.sliderHigh, this.sliderLow),
        }
      : this.sliderHigh;
    this._change(value);
    this.change.emit(value);
  }

  moveForward(key: string, pixelMove: number, limit: number) {
    const newPosition =
      this[key] + pixelMove <= limit ? this[key] + pixelMove : limit;

    this[key] = newPosition;
    this.returnCurrentValues();
  }

  moveBack(key: string, pixelMove: number, limit: number) {
    const newPosition =
      this[key] - pixelMove >= limit ? this[key] - pixelMove : limit;

    this[key] = newPosition;
    this.returnCurrentValues();
  }

  onSliderMove(key: string, position: SliderPointerChange) {
    if (this.disabled) {
      return;
    }
    const limit = this.getLimit(key, position.direction);
    const pixelMove = this.getSteps(position) * this.step;

    position.direction === 1
      ? this.moveForward(key, pixelMove, limit)
      : this.moveBack(key, pixelMove, limit);
  }

  getSelectionWidth() {
    const baseValue = Number.isInteger(this.sliderLow)
      ? this.sliderLow
      : this.min;
    this.selectionLeft =
      ((Math.min(baseValue, this.sliderHigh) - this.min) / this.max) * 100;
    this.selectionWidth =
      (Math.abs(this.sliderHigh - baseValue) / this.max) * 100;
  }
}
