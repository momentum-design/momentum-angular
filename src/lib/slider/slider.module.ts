import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SliderComponent } from './slider.component';
import { SliderPointerComponent } from './slider-pointer.component';

@NgModule({
  declarations: [SliderComponent, SliderPointerComponent],
  imports: [CommonModule],
  exports: [SliderComponent],
})
export class SliderModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: SliderModule,
      providers: [],
    };
  }
}
