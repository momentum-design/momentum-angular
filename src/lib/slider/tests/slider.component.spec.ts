import { Component, ViewChild } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { SliderChange, SliderComponent } from '../slider.component';
import { SliderPointerChange } from '../slider-pointer.component';
import { SliderModule } from '../slider.module';

describe('SliderComponent', () => {
  let testComponent: TestAppComponent;
  let fixture: ComponentFixture<TestAppComponent>;

  @Component({
    selector: 'test-app',
    template: `
      <cui-slider
        [min]="0"
        [max]="100"
        (change)="change($event)"
      ></cui-slider>
    `,
  })
  class TestAppComponent {
    @ViewChild(SliderComponent) public slider: SliderComponent;

    change(_: SliderChange) {}
  }

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestAppComponent],
      imports: [FormsModule, SliderModule],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestAppComponent);
    testComponent = fixture.componentInstance;
  });

  it('should create and match snapshot', () => {
    fixture.detectChanges();
    expect(testComponent).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should render Slider properly', () => {
    testComponent.slider.value = { low: 10, high: 40 };
    fixture.detectChanges();
    const slider = fixture.nativeElement.querySelector('cui-slider');
    expect(slider.querySelectorAll('.cui-slider__pointer').length).toEqual(2);
    expect(slider.querySelectorAll('.cui-slider__bar').length).toEqual(1);
    expect(slider.querySelectorAll('.cui-slider__selection').length).toEqual(1);
  });

  it('when only one sliderPointer is present', () => {
    testComponent.slider.value = 40;
    fixture.detectChanges();
    const slider = fixture.nativeElement.querySelector('cui-slider');
    expect(slider.querySelectorAll('.cui-slider__pointer').length).toEqual(1);
  });

  it('when slider moves and attempts to cross over', () => {
    testComponent.slider.value = { low: 10, high: 40 };
    testComponent.slider.getSliderWidth = () => 100;
    fixture.detectChanges();

    spyOn(testComponent, 'change');
    testComponent.slider.onSliderMove('sliderLow', <SliderPointerChange>{
      from: 10,
      to: 70,
      direction: 1,
    });
    fixture.whenStable().then(() => {
      expect(testComponent.change).toHaveBeenCalledWith({
        low: 40,
        high: 40,
      });
    });
    expect(testComponent.slider.sliderLow).toEqual(40);
    expect(testComponent.slider.sliderHigh).toEqual(40);
  });

  it('should slider stops at boundaries when moving in forward direction', () => {
    testComponent.slider.value = { low: 10, high: 40 };
    testComponent.slider.getSliderWidth = () => 100;
    fixture.detectChanges();

    spyOn(testComponent, 'change');
    testComponent.slider.onSliderMove('sliderHigh', <SliderPointerChange>{
      from: 40,
      to: 110,
      direction: 1,
    });
    fixture.whenStable().then(() => {
      expect(testComponent.change).toHaveBeenCalledWith({
        low: 10,
        high: 100,
      });
    });
    expect(testComponent.slider.sliderHigh).toEqual(100);
  });

  it('should slider stops at boundaries when moving in backward direction', () => {
    testComponent.slider.value = { low: 10, high: 40 };
    testComponent.slider.getSliderWidth = () => 100;
    fixture.detectChanges();

    spyOn(testComponent, 'change');
    testComponent.slider.onSliderMove('sliderLow', <SliderPointerChange>{
      from: 10,
      to: -20,
      direction: -1,
    });
    fixture.whenStable().then(() => {
      expect(testComponent.change).toHaveBeenCalledWith({ low: 0, high: 40 });
    });
    expect(testComponent.slider.sliderLow).toEqual(0);
  });

  it('on keyboard events arrow right', () => {
    testComponent.slider.value = 40;
    testComponent.slider.getSliderWidth = () => 100;
    fixture.detectChanges();

    const slider = fixture.nativeElement.querySelector('cui-slider');

    slider.querySelector('.cui-slider__pointer').dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowRight',
      })
    );
    expect(testComponent.slider.sliderHigh).toEqual(41);
  });

  it('on keyboard events arrow left', () => {
    testComponent.slider.value = 40;
    testComponent.slider.getSliderWidth = () => 100;
    fixture.detectChanges();

    const slider = fixture.nativeElement.querySelector('cui-slider');

    slider.querySelector('.cui-slider__pointer').dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowLeft',
      })
    );
    expect(testComponent.slider.sliderHigh).toEqual(39);
  });

  it('on keyboard events arrow up', () => {
    testComponent.slider.value = 40;
    testComponent.slider.getSliderWidth = () => 100;
    fixture.detectChanges();

    const slider = fixture.nativeElement.querySelector('cui-slider');

    slider.querySelector('.cui-slider__pointer').dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowUp',
      })
    );
    expect(testComponent.slider.sliderHigh).toEqual(41);
  });

  it('on keyboard events arrow down', () => {
    testComponent.slider.value = 40;
    testComponent.slider.getSliderWidth = () => 100;
    fixture.detectChanges();

    const slider = fixture.nativeElement.querySelector('cui-slider');

    slider.querySelector('.cui-slider__pointer').dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowDown',
      })
    );
    expect(testComponent.slider.sliderHigh).toEqual(39);
  });
});
