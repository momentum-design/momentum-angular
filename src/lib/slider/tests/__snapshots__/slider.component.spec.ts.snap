// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SliderComponent should create and match snapshot 1`] = `
<test-app
  slider={[Function SliderComponent]}
>
  <md-slider
    class="md-slider"
    ng-reflect-max="100"
    ng-reflect-min="0"
  >
    <span
      class="md-slider__bar"
    />
    <span
      class="md-slider__selection"
      ng-reflect-ng-style="[object Object]"
    />
    
    <md-slider-pointer
      class="md-slider__pointer"
      ng-reflect-position="NaN"
      role="button"
      tabindex="0"
    />
    <div>
      
      <span
        class="md-slider__hashlabel"
      >
        0
      </span>
      <span
        class="md-slider__hashlabel"
      >
        100
      </span>
    </div>
  </md-slider>
</test-app>
`;
